tests/components/dyson/test_fan.py

```diff
@@ -320,14 +384,542 @@ def test_service_set_night_mode(self):
         self.hass.data[DYSON_DEVICES] = []
         dyson_device.entity_id = 'fan.living_room'
         self.hass.data[dyson.DYSON_FAN_DEVICES] = [dyson_device]
-        dyson.setup_platform(self.hass, None, mock.MagicMock())
+        dyson.setup_platform(self.hass, None,
+                             mock.MagicMock(), mock.MagicMock())
 
         self.hass.services.call(dyson.DOMAIN, dyson.SERVICE_SET_NIGHT_MODE,
                                 {"entity_id": "fan.bed_room",
                                  "night_mode": True}, True)
-        assert not dyson_device.night_mode.called
+        assert not dyson_device.set_night_mode.called
 
         self.hass.services.call(dyson.DOMAIN, dyson.SERVICE_SET_NIGHT_MODE,
                                 {"entity_id": "fan.living_room",
                                  "night_mode": True}, True)
-        dyson_device.night_mode.assert_called_with(True)
+        dyson_device.set_night_mode.assert_called_with(True)
+
+
+@mock.patch('libpurecool.dyson.DysonAccount.login', return_value=True)
```

Comment:
The regular moch patch doesn't work for coroutine tests. Please use the patch from asynctest instead.

Reply:
Thanks for pointing this out, it does fail in quite an insidious way, because the asserts always return true! 

Now I tried asynctest.patch, but the problem I'm running into is that the entity's service method never gets called. Looking a bit further into it, it seems that the hass's services.call method is a mock, so the test written like this won't work at all(I can test if the services.call method was called, but thats not really useful). I didn't have the time to go deeper into how the hass test instance is created tho, I'm probably misunderstanding something there. I'm going to check it now. 

In the meantime it would be great if you can point me out in the general direction how best to test the configuration of custom services.

Reply:
Mock the library method that the entity method calls, which in turn the service calls. Then it's just to set up the component that sets up the correct platform which adds an entity. Then call the service and check that the mocked library method is called correctly.

Reply:
Thank you, I just realised what's going on.

I completely misunderstood how the pytest fixtures work so by declaring the test method like this:
```python
@asynctest.patch('libpurecool.dyson.DysonAccount.login', return_value=True)
@asynctest.patch('libpurecool.dyson.DysonAccount.devices',
            return_value=[_get_dyson_purecool_device()])
async def test_purecool_service_set_night_mode(hass, mocked_devices):
```
I was actually getting the return value of the second patch as hass and the return value of the first patch as mocked_devices. So when looking at it in the debugger the hass object looked like a mock...

Rewriting it like this works fine: 
```python
@asynctest.patch('libpurecool.dyson.DysonAccount.login', return_value=True)
@asynctest.patch('libpurecool.dyson.DysonAccount.devices',
                 return_value=[_get_dyson_purecool_device()])
async def test_purecool_service_set_night_mode(devices, login, hass):
    """Test set night mode service."""
    dyson_parent.setup(hass, _get_config())
    hass.async_block_till_done()
```
I'll rewrite all the tests now so that they assert that the mocked library methods are called instead of the entity. Thanks for the tip.